keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,362 @@
+
+#### 'kubectl get endpoints' Command Display for Dual-Stack Backend Pods
+There should be no changes required for the 'kubectl get endpoints ...' command display. The command output should include both IPv4 and IPv6 addresses as individual endpoints, e.g.:
+```
+       kube-master# kubectl get endpoints
+       NAME         ENDPOINTS                         AGE
+       kubernetes   [fd00:90::2]:6443,10.0.0.2:6643   20m
+       kube-master#
+```
+
+### kube-proxy Operation
+
+In order to support dual-stack Kubernetes services, kube-proxy will be modified to support parallel support for iptables and ip6tables mappings. This will require the implementation of a second "proxier" interface in the Kube-Proxy server in order to modify and track changes to iptables and ip6tables in parallel.
+
+#### Kube-Proxy Startup Configuration Changes
+
+##### Multiple bind addresses configuration
+New [kube-proxy configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/) arguments will be added to allow a user to specify separate IPv4 and IPv6 bind addresses for the kube-proxy server:
+```
+- --bind-address string     [Existing argument, currently used for IPv4 OR IPv6, will now be IPv4-only. Default: 0.0.0.0]
+- --bind-address-v6 string  [New argument. Default: '::']
+- --healthz-bind-address string     [Existing argument, currently used for IPv4 OR IPv6, will now be IPv4-only. Default: 0.0.0.0]
+- --healthz-bind-address-v6 string  [New argument. Default: '::']
+- --metrics-bind-address string     [Existing argument, currently used for IPv4 OR IPv6, will now be IPv4-only. Default: 0.0.0.0]
+- --metrics-bind-address-v6 string  [New argument. Default: '::']
+```
+
+##### Multiple cluster CIDRs configuration
+A new [kube-proxy configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/) argument will be added to allow a user to specify multiple cluster CIDRs.
+- --cluster-cidr  string       [Existing argument. Retained for backwords compatibility. Can be an IPv4 or IPv6 CIDR]
+- --cluster-cidrs stringSlice  [New argument. Comma-separated list of CIDRs]
+If both arguments are specified, the CIDRs from both will be combined.
+
+### IPVS Support and Operation
+
+\<TBD\>
+
+### Support of Health/Liveness/Readiness Probes
+
+Currently, health, liveness, and readiness probes are defined without any concern for IP addresses or families. For the first release of dual-stack support, no configuration "knobs" will be added for probe definitions. A probe for a dual-stack pod will be deemed successful if either an IPv4 or IPv6 response is received. (QUESTION: Does the current probe implementation include DNS lookups, or are IP addresses hard coded?)
```

Comment:
> Does the current probe implementation include DNS lookups, or are IP addresses hard coded

I believe you can configure a `host` for the probe which will be resolved via DNS if it is a DNS name rather than an exact IP. [See here](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#configure-probes)